src/i2s.rs

```diff
@@ -0,0 +1,1835 @@
+
+        // Initialize the interal side of the driver.
+
+        // Safety: We know that internal is properly aligned and non-null.
+        // We are initializing i2s here.
+        unsafe {
+            (*internal).i2s = port as u8;
+        }
+
+        let ll_config: i2s_chan_config_t = config.as_sdk(port);
+
+        // Safety: We initialize tx_chan_handle and rx_chan_handle to null where necessary, and return _pointers_ to
+        // them, not the uninitialized values, otherwise.
+        let (rx_p, tx_p): (*mut i2s_chan_handle_t, *mut i2s_chan_handle_t) = unsafe {
+            match config.channels {
+                config::ChannelOpen::Both => (
+                    &mut (*internal).rx_chan_handle,
+                    &mut (*internal).tx_chan_handle,
+                ),
+                config::ChannelOpen::Rx => {
+                    (*internal).tx_chan_handle = null_mut();
+                    (&mut (*internal).rx_chan_handle, null_mut())
+                }
+                config::ChannelOpen::Tx => {
+                    (*internal).rx_chan_handle = null_mut();
+                    (null_mut(), &mut (*internal).tx_chan_handle)
+                }
+            }
+        };
+
+        // We're done with config, so move it into the internal driver struct.
+        unsafe {
+            (*internal).config = config;
+        }
+
+        // Safety: &ll_config is a valid pointer to an i2s_chan_config_t. rx_p and tx_p are either valid pointers to
+        // the internal.rx_chan_handle and internal.tx_chan_handle, or null.
+        unsafe { esp!(i2s_new_channel(&ll_config, tx_p, rx_p))? };
+
+        // At this point, everything except the available atomics is initialized. Initialize the atomics with Release
+        // ordering to ensure the above values are visible to other threads.
+        unsafe {
+            (*internal)
+                .rx_chan_available
+                .store(!rx_p.is_null(), Ordering::Release);
+        }
+        unsafe {
+            (*internal)
+                .tx_chan_available
+                .store(!tx_p.is_null(), Ordering::Release);
+        }
+
+        // Safety: Box would drop the internal driver (which is statically allocated), but we leak the box by wrapping
+        // it in a ManuallyDrop that we never drop.
+        let internal = ManuallyDrop::new(unsafe { Box::from_raw(internal) });
+
+        Ok(Self {
+            internal,
+            _p: PhantomData,
+        })
+    }
+
+    /// Open the receive channel.
+    ///
+    /// The returned channel will be in the `READY` state: initialized but not yet started.
+    ///
+    /// # Errors
+    /// This will return an [EspError] with `ESP_ERR_NOT_FOUND` if the channel is already open or the I2S peripheral
+    /// was not configured to receive data.
+    ///
+    /// # TODO
+    /// This currenly only supports the standard configuration mode.
+    pub fn open_rx_channel<'c, 'cb, C, Mclk, Bclk, Ws, Dout, Din>(
```

Comment:
As per above, perhaps we don't need this complexity, even if ESP IDF models it in such a complex way? For me, the thing that ruins the approach is that you anyway have to specify - at driver construction time - whether it will operate in TX, RX or Both mode. Then why not pass all the pins at construction time?

As for passing pins that you won't use - e.g. RX pins for TX mode-only, you can parameterize the driver by a typestate of structs - `Duplex` `Rx` `Tx` and then expose different `new` constructors on `I2sDriver<Duplex>` vs `I2sDriver<Tx>` and so on.

You can also define `write` directly on the driver then, but only when the typestate struct is `Tx` or `Duplex`. You can even define marker traits - e.g. `TxSupported` or similar and make `Tx` and `Duplex` implement it to avoid having to define `write` twice for `Tx` and for `Duplex`.

Yes, that's one extra generic, but only one and you can always type alias e.g. `pub type I2sTxDriver<'d> = I2sDriver<Duplex, 'd>`. 

And this way completely get away from the notion of `I2sChannel`.

Reply:
I'll have to see if we can do this through a low-power event. IIRC ESP-IDF requires us to close the channels, which on our side would naively mean dropping the driver—and, thus, dropping the pins.

But we don't have to be naive about this; we could hold a separate state internally, keeping track of whether the channel is open and allowing us to reopen it. It would add complexity, but everything about I2S just screams complex, anyway. <sigh>

Reply:
Alright, this has been done in the refactoring.